# This script contains:
1. Importing Libraries and Data
2. Dropping Column
3. Renaming Columns
4. Changing a Variable's Data Type
5. Transposing Data
6. Adding Index
7. Data Dictionary
8. Subsetting
9. Question 2
10. Question 3
11. Question 4
12. Question 5
13. Question 6
14. QUestion 7
15. Question 8
16. Question 9
17. Data Export

In [1]:
# Importing Libraries and Data
import pandas as pd
import numpy as np
import os

In [3]:
path= r'/Users/navneetjha/Desktop/Career foundry Docs/exercises/Data Immersion /Achievement 4/INSTACART BASKET ANALYSIS JULY14-2024'

In [4]:
path

'/Users/navneetjha/Desktop/Career foundry Docs/exercises/Data Immersion /Achievement 4/INSTACART BASKET ANALYSIS JULY14-2024'

In [5]:
df_prods=pd.read_csv(os.path.join(path, '02 DATA', 'ORIGINAL', 'Raw', 'products.csv'), index_col = False)

In [6]:
df_ords=pd.read_csv(os.path.join(path, '02 DATA', 'ORIGINAL', 'Raw', 'orders.csv'), index_col = False)

In [7]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 2. Dropping Column

In [8]:
df_ords.drop(columns=['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [10]:
df_ords_2=df_ords.drop(columns=['eval_set'])

In [11]:
df_ords_2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [12]:
df_ords['days_since_prior_order'].value_counts(dropna=False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

# 3. Renaming Columns

In [14]:
df_ords_2.rename(columns={'order_dow':'order_day_of_week'},inplace=True)

In [15]:
df_ords_2.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 4. Changing a Variable's Data Type

In [17]:
df_ords_2['order_id']=df_ords_2['order_id'].astype('str')

In [18]:
df_ords_2['order_id'].dtype

dtype('O')

# 5. Transposing Data( Meaning changing row to column and vice-versa)

In [20]:
df_dep= pd.read_csv(os.path.join(path,'02 DATA', 'ORIGINAL', 'Raw', 'departments.csv'), index_col = False)

In [21]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [22]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
df_dep_t=df_dep.T

In [24]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 6. Adding Index to df_dep_t

In [26]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [27]:
# Taking the first row of df_dep_t as the header
new_header=df_dep_t.iloc[0]

In [28]:
new_header

0    department
Name: department_id, dtype: object

In [29]:
# Taking the data under the header row as new data frame 
df_dep_t_new=df_dep_t[1:]

In [30]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [31]:
df_dep_t_new.columns=new_header   #set the header row as the df header

In [32]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 7. Data Dictionary (Command use is to_dict)

In [34]:
data_dict=df_dep_t_new.to_dict('index') # index is used to tell python use the numbered rows as the key values for the entries in the dictionary.

In [35]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [36]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [37]:
data_dict.get('19')

{'department': 'snacks'}

# 8. Subsetting --- Searching for products in the same department

In [39]:
df_snacks=df_prods[df_prods['department_id']==19]

In [40]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [41]:
df_prods.head(40)

product_id                                       product_name  aisle_id  \
0            1                         Chocolate Sandwich Cookies        61   
1            2                                   All-Seasons Salt       104   
2            3               Robust Golden Unsweetened Oolong Tea        94   
3            4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4            5                          Green Chile Anytime Sauce         5   
5            6                                       Dry Nose Oil        11   
6            7                     Pure Coconut Water With Orange        98   
7            8                  Cut Russet Potatoes Steam N' Mash       116   
8            9                  Light Strawberry Blueberry Yogurt       120   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
11          12                         Chocolate Fudge Layer Cake       119   
12          13                                  Saline Nasal Mist        11   
13          14                     Fresh Scent Dishwasher Cleaner        74   
14          15                           Overnight Diapers Size 6        56   
15          16                      Mint Chocolate Flavored Syrup       103   
16          17                                  Rendered Duck Fat        35   
17          18                Pizza for One Suprema  Frozen Pizza        79   
18          19   Gluten Free Quinoa Three Cheese & Mushroom Blend        63   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
20          21                   Small & Medium Dental Dog Treats        40   
21          22                  Fresh Breath Oral Rinse Mild Mint        20   
22          23                             Organic Turkey Burgers        49   
23          24  Tri-Vi-Sol® Vitamins A-C-and D Supplement Drop...        47   
24          25            Salted Caramel Lean Protein & Fiber Bar         3   
25          26        Fancy Feast Trout Feast Flaked Wet Cat Food        41   
26          27  Complete Spring Water Foaming Antibacterial Ha...       127   
27          28                                  Wheat Chex Cereal       121   
28          29  Fresh Cut Golden Sweet No Salt Added Whole Ker...        81   
29          30         Three Cheese Ziti, Marinara with Meatballs        38   
30          31                                 White Pearl Onions       123   
31          32                      Nacho Cheese White Bean Chips       107   
32          33                      Organic Spaghetti Style Pasta       131   
33          34                                                NaN       121   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
35          36  Traditional Lasagna with Meat Sauce Savory Ita...        38   
36          37                 Noodle Soup Mix With Chicken Broth        69   
37          38                    Ultra Antibacterial Dish Liquid       100   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   

    department_id  prices  
0              19     5.8  
1              13     9.3  
2               7     4.5  
3               1    10.5  
4              13     4.3  
5              11     2.6  
6               7     4.4  
7               1     1.1  
8              16     7.0  
9               7     8.4  
10              7     2.8  
11              1     9.4  
12             11     1.1  
13             17     6.5  
14             18    11.2  
15             19     5.2  
16             12    17.1  
17              1    10.6  
18              9    12.0  
19              7     6.0  
20              8    10.0  
21             11     9.5  
22             12     8.2  
23             11     9.1  
24             19     1.9  
25              8    11.3  
26      

In [42]:
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19] # we can replace ==19 with .isin ([19]) too. 

In [43]:
df_snacks_2

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

TASK 4.4

# 9. Question 2

In [132]:
df_ords_2.head(10)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   
5  3367565       1             6                  2                  7   
6   550135       1             7                  1                  9   
7  3108588       1             8                  1                 14   
8  2295261       1             9                  1                 16   
9  2550362       1            10                  4                  8   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0

In [46]:
df_ords_2['user_id']=df_ords_2['user_id'].astype('str') # changing variable user_id into string

In [47]:
df_ords_2['user_id'].dtype

dtype('O')

The data type of user_id shows 'o' which means an object in Panda.

# 10.Question 3 (Renaming Column)

In [88]:
df_ords_2['user_id'].dtype

dtype('O')

In [90]:
df_ords_4 = df_ords_2.rename(columns={'days_since_prior_order':'days_since_previous_order'},inplace=False)

In [92]:
df_ords_2.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [94]:
df_ords_4.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_previous_order  
0                        NaN  
1                       15.0  
2                       21.0  
3                       29.0  
4                       28.0

# 11. Question 4 (Busiest Hour of the day)

In [165]:
df_ords_2['order_hour_of_day'].value_counts(dropna= False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

Busiest hour is 10AM

# 12. Question 5 Value of 4 in the "department_id"

In [97]:
print(data_dict.get('4'))

{'department': 'produce'}


# 13. Question 6 Breakfast Items Sales

In [99]:
df_breakfast_items=df_prods[df_prods['department_id']==14]

In [103]:
df_breakfast_items.head() #Printing first five rows of breakfast items dataframe

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# 14. Question 7 Products for Dinner Parties

In [105]:
df_dinner_parties_items=df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [107]:
df_dinner_parties_items.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

# 15. Question 8. Finding number of rows of df_dinner_parties_items.

In [109]:
df_dinner_parties_items.shape

(7650, 5)

There are 7650 rows and 5 columns in the dinner parties dataframe

# 16. Question 9. Information about user_id of 1

In [111]:
df_user_id_1_information=df_ords[df_ords['user_id']==1] 

In [119]:
df_user_id_1_information.head() # Printing first five rows of the df_user_id_1_information.

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [121]:
df_user_id_1_information.describe() # Extracting descriptive information

order_id  user_id  order_number  order_dow  order_hour_of_day  \
count  1.100000e+01     11.0     11.000000  11.000000          11.000000   
mean   1.923450e+06      1.0      6.000000   2.636364          10.090909   
std    1.071950e+06      0.0      3.316625   1.286291           3.477198   
min    4.315340e+05      1.0      1.000000   1.000000           7.000000   
25%    8.690170e+05      1.0      3.500000   1.500000           7.500000   
50%    2.295261e+06      1.0      6.000000   3.000000           8.000000   
75%    2.544846e+06      1.0      8.500000   4.000000          13.000000   
max    3.367565e+06      1.0     11.000000   4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

The customer with user_id 1 shopping behavior analysis:
* There are 11 counts of entry about user_id in all columns except in days_since_prior_order. It is missing one entry in days_since_prior_order column.
* Shopping Day Window = Sunday to Wednesday.
* No Shopping Day     =    Thrusday to Saturday.
* Shopping Time Window = 7AM-4PM.
* No Shopping Time = 4:01PM-6:59AM
* Average shopping time window is around 10AM.
* Maximum Day since previous order = 30
* Average day from prior ordder = 19. Meaning the user on an average places order on every 19th day from last shopping day.
* The standard deviation for day of order is approximately 1 day.
* The standard deviation for order_hour_of_day in a little more that 3 hours.
* The standard deviation for days_since_prior_order is 9 days.

# 17. Data Export

In [138]:
df_ords.to_csv(os.path.join(path,'02 DATA','PREPARED DATA', 'orders_wrangled.csv'))

In [140]:
df_prods.to_csv(os.path.join(path,'02 DATA','PREPARED DATA', 'products_wrangled.csv'))

In [142]:
df_dep_t_new.to_csv(os.path.join(path,'02 DATA','PREPARED DATA','departments_wrangled.csv'))